map/framework.cpp

```diff
@@ -1650,11 +1650,52 @@ void Framework::ConnectToGpsTracker()
 void Framework::DisconnectFromGpsTracker()
 {
   m_connectToGpsTrack = false;
-  GpsTracker::Instance().Disconnect();
+  auto & tracker = GpsTracker::Instance();
+  tracker.Disconnect();
+  tracker.SetEnabled(false);
   if (m_drapeEngine)
     m_drapeEngine->ClearGpsTrackPoints();
 }
 
+void Framework::StartTrackRecording()
+{
+  auto & tracker = GpsTracker::Instance();
+  if (!tracker.IsEnabled())
+    tracker.SetEnabled(true);
+  m_connectToGpsTrack = true;
+  if (m_drapeEngine)
+  {
+    m_drapeEngine->ClearGpsTrackPoints();
+    tracker.Connect(bind(&Framework::OnUpdateGpsTrackPointsCallback, this, _1, _2));
+  }
+}
+
+void Framework::StopTrackRecording()
+{
+  m_connectToGpsTrack = false;
+  auto & tracker = GpsTracker::Instance();
+  tracker.Disconnect();
+  tracker.SetEnabled(false);
+  if (m_drapeEngine)
+    m_drapeEngine->ClearGpsTrackPoints();
+}
+
+void Framework::StopTrackRecordingAndSaveWithName(std::string const & name)
+{
+  GetBookmarkManager().SaveTrackRecording(name);
+  StopTrackRecording();
```

Comment:
Should it be in a different order? An active recorder may start processing GPS point during save.
```suggestion
  StopTrackRecording();
  GetBookmarkManager().SaveTrackRecording(name);
```

Reply:
If (when) we decide to clean up the dotted line when the user press the "Stop without saving" the `StopTrackRecording();` will remove the recorded track before saving.

It is safer to save the track and only then make all the clean-ups.

Reply:
> An active recorder may start processing GPS point during save.

It isn't a problem because the mutex is used for the getter
```
std::vector<location::GpsTrackInfo> GpsTrack::GetCurrentTrack() const
{
  lock_guard<mutex> lg(m_dataGuard);
...
```

Reply:
It is better to clearly separate saving the track and removing it from the screen, e.g. by using different methods.

Reply:
Does a mutex mean that for very large tracks (e.g. weeks) the main (location) thread will be blocked on that mutex until the track is saved?